File: src/julia_cfunctions.jl
const julia_sind = @cfunction(sind, Cdouble, (Cdouble,))

# Mutually recursive fibonacci implementation which bounces back and
# forth between Julia and C.
function mutual_fibonacci(n::Cint, callback::Ptr{Cvoid})::Cint
    if n <= 1
        return n
    end

    return ccall(callback, Cint, (Cint,), n - 1) +
           ccall(callback, Cint, (Cint,), n - 2)
end

const mutual_fibonacci_ptr = @cfunction(mutual_fibonacci, Cint, (Cint, Ptr{Cvoid}))

# Example of receiving an array of data. This function adds 1 to the
# first element, 2 to the second, and so on.
function add_element_number(x::Ptr, n::Integer)
    for i = 1:n
        element = unsafe_load(x, i)
        unsafe_store!(x, element + i, i)
    end
    return
end

# @cfunction can instantiate pointers to multiple specializations of
# the function for different argument types.
const add_element_number_int = @cfunction(add_element_number, Cvoid,
                                          (Ptr{Cint}, Cint))
const add_element_number_float = @cfunction(add_element_number, Cvoid,
                                            (Ptr{Cfloat}, Cint))

File: c/julia_cfunctions.c
#include "julia_embedding.h"
#include "julia_cfunctions.h"

double (*julia_sind)(double x);
int (*mutual_fibonacci)(int n, int (*callback)(int n));
void (*add_element_number_int)(int *x, int n);
void (*add_element_number_float)(float *x, int n);

int load_julia_cfunctions()
{
    julia_sind = get_cfunction_pointer("julia_sind");
    if (!julia_sind)
        return 0;

    mutual_fibonacci = get_cfunction_pointer("mutual_fibonacci_ptr");
    if (!mutual_fibonacci)
        return 0;

    add_element_number_int = get_cfunction_pointer("add_element_number_int");
    if (!add_element_number_int)
        return 0;

    add_element_number_float = get_cfunction_pointer("add_element_number_float");
    if (!add_element_number_float)
        return 0;

    return 1;
}
